identity_stardust/src/state_metadata_document.rs

```diff
@@ -0,0 +1,326 @@
+    }
+
+    let mut core_doc: CoreDocument = builder.build().expect("all required document fields should be set");
+
+    std::mem::swap(core_doc.also_known_as_mut(), self.0.also_known_as_mut());
+
+    StardustDocument(core_doc)
+  }
+}
+
+impl From<StardustDocument> for StateMetadataDocument {
+  /// Transforms a [`StardustDocument`] into its state metadata representation by replacing all
+  /// occurrences of its did with a placeholder.
+  fn from(mut doc: StardustDocument) -> Self {
+    let mut original_did: CoreDID = PLACEHOLDER_DID.to_owned();
+    std::mem::swap(&mut original_did, doc.0.id_mut());
+
+    let mut builder: DocumentBuilder<DIDOrPlaceholder, _, _, _> =
+      CoreDocument::builder(doc.0.properties().clone()).id(DIDOrPlaceholder::Placeholder);
+
+    let mut controllers = None;
+    std::mem::swap(&mut controllers, doc.0.controller_mut());
+
+    if let Some(controllers) = controllers {
+      for controller in controllers.into_vec().into_iter() {
+        builder = builder.controller(DIDOrPlaceholder::new(controller, &original_did));
+      }
+    }
+
+    for method in doc
+      .0
+      .verification_method_mut()
+      .as_vec_mut()
+      .drain(..)
+      .map(|method| method.map(|did| DIDOrPlaceholder::new(did, &original_did)))
+    {
+      builder = builder.verification_method(method);
+    }
+
+    let method_ref_transform = |method_ref: MethodRef| method_ref.map(|did| DIDOrPlaceholder::new(did, &original_did));
+
+    for method_ref in doc
+      .0
+      .assertion_method_mut()
+      .as_vec_mut()
+      .drain(..)
+      .map(method_ref_transform)
+    {
+      builder = builder.assertion_method(method_ref);
+    }
+    for method_ref in doc
+      .0
+      .capability_delegation_mut()
+      .as_vec_mut()
+      .drain(..)
+      .map(method_ref_transform)
+    {
+      builder = builder.capability_delegation(method_ref);
+    }
+    for method_ref in doc
+      .0
+      .capability_invocation_mut()
+      .as_vec_mut()
+      .drain(..)
+      .map(method_ref_transform)
+    {
+      builder = builder.capability_invocation(method_ref);
+    }
+    for method_ref in doc
+      .0
+      .key_agreement_mut()
+      .as_vec_mut()
+      .drain(..)
+      .map(method_ref_transform)
+    {
+      builder = builder.key_agreement(method_ref);
+    }
+    for method_ref in doc
+      .0
+      .authentication_mut()
+      .as_vec_mut()
+      .drain(..)
+      .map(method_ref_transform)
+    {
+      builder = builder.authentication(method_ref);
+    }
+
+    for service in doc.0.service_mut().as_vec_mut().drain(..) {
+      builder = builder.service(service.map(|did| DIDOrPlaceholder::new(did, &original_did)));
+    }
+
+    let mut core_doc: CoreDocument<DIDOrPlaceholder, _, _, _> =
+      builder.build().expect("all required document fields should be set");
+
+    std::mem::swap(core_doc.also_known_as_mut(), doc.0.also_known_as_mut());
+
+    StateMetadataDocument(core_doc)
+  }
```

Comment:
Same here: use the `map` function.

```suggestion
  fn from(document: StardustDocument) -> Self {
    let core_document: CoreDocument = CoreDocument::from(document);
    let id = core_document.id().clone();
    let core_doc: CoreDocument<DIDOrPlaceholder> = core_document.map(
      // Replace self-referential identifiers with a placeholder, but not others.
      // TODO: can this just be a CoreDID like PLACEHOLDER_DID instead?
      |did| {
        if did == id {
          DIDOrPlaceholder::Placeholder
        } else {
          DIDOrPlaceholder::Core(did)
        }
      },
      // Do not modify properties.
      |o| o,
    );
    StateMetadataDocument(core_doc)
  }
```

Reply:
Requires a simple `From` impl on `StardustDocument`.

```rust
impl From<StardustDocument> for CoreDocument {
  fn from(document: StardustDocument) -> Self {
    // TODO: convert any additional properties (that cannot be inferred from the Output) into
    //       JSON Object or an alternative struct in CoreDocument.
    document.0
  }
}
```